src/Plugin/resource/DataProvider/DataProviderEntity.php

```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
     return $resource_field->access($op, $interpreter);
   }
+
+  /**
+   * Get referenced ID.
+   *
+   * @param string $value
+   *   The provided value, it can be an ID or not.
+   * @param ResourceFieldInterface $resource_field
+   *   The resource field that points to another entity.
+   *
+   * @return string
+   *   If the field uses an alternate ID property, then the ID gets translated
+   *   to the original entity ID. If not, then the same provided ID is returned.
+   *
+   * @todo: Add testing to this functionality.
+   */
+  protected function getReferencedId($value, ResourceFieldInterface $resource_field) {
+    $field_definition = $resource_field->getDefinition();
+    if (empty($field_definition['referencedIdProperty'])) {
+      return $value;
+    }
+    $id_property = $field_definition['referencedIdProperty'];
+    // Get information about the the Drupal field to see what entity type we are
+    // dealing with.
+    $field_info = field_info_field($resource_field->getProperty());
+    // We support:
+    // - Entity Reference.
+    // - Taxonomy Term.
+    // - File & Image field.
+    // If you need to support other types, you can create a custom data provider
+    // that overrides this method.
+    $target_entity_type = NULL;
+    $bundles = array();
+    $query = $this->EFQObject();
+    if ($field_info['type'] == 'entityreference') {
+      $target_entity_type = $field_info['settings']['target_type'];
+      $bundles = empty($field_info['settings']['handler_settings']['target_bundles']) ? array() : $field_info['settings']['handler_settings']['target_bundles'];
+    }
+    elseif ($field_info['type'] == 'file') {
+      $target_entity_type = 'file';
+    }
+    elseif ($field_info['type'] == 'taxonomy_term_reference') {
+      $target_entity_type = 'taxonomy_term';
+      // Narrow down with the vocabulary information. Very useful if there are
+      // multiple terms with the same name in different vocabularies.
+      foreach ($field_info['settings']['allowed_values'] as $allowed_value) {
+        $bundles[] = $allowed_value['vocabulary'];
+      }
+    }
+    if (empty($target_entity_type)) {
+      return $value;
+    }
+
+    // Now we have the entity type and bundles to look for the entity based on
+    // the contents of the field or the entity property.
+    $query->entityCondition('entity_type', $target_entity_type);
+    if (!empty($bundles)) {
+      // Narrow down for bundles.
+      $query->entityCondition('bundle', $bundles, 'IN');
+    }
+    // Check if the $id_property is a field or a property.
+    if (field_info_field($id_property)) {
+      $query->fieldCondition($id_property, $resource_field->getColumn(), $value);
+    }
+    else {
+      $query->propertyCondition($id_property, $value);
+    }
+    // Only one result is returned. Make sure that you are using "unique" fields
```

Comment:
Make sure that you are using "unique" fields .. => maybe: `This assumes the reference fields are unique for every entity.`